rosbag2_compression/src/rosbag2_compression/sequential_compression_reader.cpp

```diff
@@ -0,0 +1,210 @@
+{
+  compression_mode_ = rosbag2_compression::compression_mode_from_string(metadata_.compression_mode);
+  if (compression_mode_ != rosbag2_compression::CompressionMode::NONE) {
+    if (metadata_.compression_format == "zstd") {
+      decompressor_ = std::make_unique<rosbag2_compression::ZstdDecompressor>();
+      // Decompress the first file so that it is readable.
+      ROSBAG2_COMPRESSION_LOG_DEBUG_STREAM("Decompressing " << get_current_file().c_str());
+      set_current_file(decompressor_->decompress_uri(get_current_file()));
+    } else {
+      std::stringstream err;
+      err << "Unsupported compression format " << metadata_.compression_format;
+      throw std::invalid_argument{err.str()};
+    }
+  } else {
+    throw std::invalid_argument{
+            "SequentialCompressionReader requires a CompressionMode that is not NONE!"};
+  }
+}
+
+void SequentialCompressionReader::open(
+  const rosbag2_cpp::StorageOptions & storage_options,
+  const rosbag2_cpp::ConverterOptions & converter_options)
+{
+  if (metadata_io_->metadata_file_exists(storage_options.uri)) {
+    metadata_ = metadata_io_->read_metadata(storage_options.uri);
+    if (metadata_.relative_file_paths.empty()) {
+      ROSBAG2_COMPRESSION_LOG_WARN("No file paths were found in metadata.");
+      return;
+    }
+    file_paths_ = metadata_.relative_file_paths;
+    current_file_iterator_ = file_paths_.begin();
+    setup_decompression();
+
+    storage_ = storage_factory_->open_read_only(
+      *current_file_iterator_, metadata_.storage_identifier);
+    if (!storage_) {
+      throw std::runtime_error{"No storage could be initialized. Abort"};
+    }
+  } else {
+    throw std::runtime_error{"Compression is not supported for legacy bag files."};
+  }
+  const auto & topics = metadata_.topics_with_message_count;
+  if (topics.empty()) {
+    ROSBAG2_COMPRESSION_LOG_WARN("No topics were listed in metadata.");
+    return;
+  }
+
+  // Currently a bag file can only be played if all topics have the same serialization format.
+  check_topics_serialization_formats(topics);
+  check_converter_serialization_format(
+    converter_options.output_serialization_format,
+    topics[0].topic_metadata.serialization_format);
+}
+
+bool SequentialCompressionReader::has_next()
+{
+  if (storage_) {
+    // If there's no new message, check if there's at least another file to read and update storage
+    // to read from there. Otherwise, check if there's another message.
+    if (!storage_->has_next() && has_next_file()) {
+      load_next_file();
+      storage_ = storage_factory_->open_read_only(
+        *current_file_iterator_, metadata_.storage_identifier);
+    }
+    return storage_->has_next();
+  }
+  throw std::runtime_error{"Bag is not open. Call open() before reading."};
+}
+
+std::shared_ptr<rosbag2_storage::SerializedBagMessage> SequentialCompressionReader::read_next()
+{
+  if (storage_) {
+    assert(decompressor_);
```

Comment:
asserts won't be triggered when being compiled in release mode. Why not throwing here as well when `decompressor_` is null just like the storage?

Reply:
`assert` was chosen since `decompressor_` cannot be null; `SequentialCompressionReader` should have already thrown an exception.

I can change it to throw an exception